File: tests/rosetta/transpiler/php/determine-if-a-string-is-squeezable.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
$__start_mem = memory_get_usage();
$__start = _now();
  function padLeft($n, $width) {
  global $strings, $chars, $i, $j, $c, $ss;
  $s = _str($n);
  while (strlen($s) < $width) {
  $s = ' ' . $s;
};
  return $s;
};
  function squeeze($s, $ch) {
  global $strings, $chars, $j, $ss;
  $out = '';
  $prev = false;
  $i = 0;
  while ($i < strlen($s)) {
  $c = substr($s, $i, $i + 1 - $i);
  if ($c == $ch) {
  if (!$prev) {
  $out = $out . $c;
  $prev = true;
};
} else {
  $out = $out . $c;
  $prev = false;
}
  $i = $i + 1;
};
  return $out;
};
  $strings = ['', '"If I were two-faced, would I be wearing this one?" --- Abraham Lincoln ', '..1111111111111111111111111111111111111111111111111111111111111117777888', 'I never give \'em hell, I just tell the truth, and they think it\'s hell. ', '                                                   ---  Harry S Truman  ', 'The better the 4-wheel drive, the further you\'ll be from help when ya get stuck!', 'headmistressship', 'aardvark', '😍😀🙌💃😍😍😍🙌'];
  $chars = [[' '], ['-'], ['7'], ['.'], [' ', '-', 'r'], ['e'], ['s'], ['a'], ['😍']];
  $i = 0;
  while ($i < count($strings)) {
  $j = 0;
  $s = $strings[$i];
  while ($j < count($chars[$i])) {
  $c = $chars[$i][$j];
  $ss = squeeze($s, $c);
  echo rtrim('specified character = \'' . $c . '\''), PHP_EOL;
  echo rtrim('original : length = ' . padLeft(strlen($s), 2) . ', string = «««' . $s . '»»»'), PHP_EOL;
  echo rtrim('squeezed : length = ' . padLeft(strlen($ss), 2) . ', string = «««' . $ss . '»»»'), PHP_EOL;
  echo rtrim(''), PHP_EOL;
  $j = $j + 1;
};
  $i = $i + 1;
}
$__end = _now();
$__end_mem = memory_get_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/rosetta/transpiler/FS/determine-if-a-string-is-squeezable.fs
// Generated 2025-07-31 00:10 +0700

exception Return

let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _substring (s:string) (start:int) (finish:int) =
    let len = String.length s
    let mutable st = if start < 0 then len + start else start
    let mutable en = if finish < 0 then len + finish else finish
    if st < 0 then st <- 0
    if st > len then st <- len
    if en > len then en <- len
    if st > en then st <- en
    s.Substring(st, en - st)

let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec padLeft (n: int) (width: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable n = n
    let mutable width = width
    try
        let mutable s: string = string n
        while (String.length s) < width do
            s <- " " + s
        __ret <- s
        raise Return
        __ret
    with
        | Return -> __ret
let rec squeeze (s: string) (ch: string) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable s = s
    let mutable ch = ch
    try
        let mutable out: string = ""
        let mutable prev: bool = false
        let mutable i: int = 0
        while i < (String.length s) do
            let c: string = _substring s i (i + 1)
            if c = ch then
                if not prev then
                    out <- out + c
                    prev <- true
            else
                out <- out + c
                prev <- false
            i <- i + 1
        __ret <- out
        raise Return
        __ret
    with
        | Return -> __ret
let strings: string array = [|""; "\"If I were two-faced, would I be wearing this one?\" --- Abraham Lincoln "; "..1111111111111111111111111111111111111111111111111111111111111117777888"; "I never give 'em hell, I just tell the truth, and they think it's hell. "; "                                                   ---  Harry S Truman  "; "The better the 4-wheel drive, the further you'll be from help when ya get stuck!"; "headmistressship"; "aardvark"; "😍😀🙌💃😍😍😍🙌"|]
let chars: string array array = [|[|" "|]; [|"-"|]; [|"7"|]; [|"."|]; [|" "; "-"; "r"|]; [|"e"|]; [|"s"|]; [|"a"|]; [|"😍"|]|]
let mutable i: int = 0
while i < (Seq.length strings) do
    let mutable j: int = 0
    let mutable s: string = strings.[i]
    while j < (Seq.length (chars.[i])) do
        let c: string = (chars.[i]).[j]
        let ss: string = squeeze s c
        printfn "%s" (("specified character = '" + c) + "'")
        printfn "%s" (((("original : length = " + (unbox<string> (padLeft (String.length s) 2))) + ", string = «««") + s) + "»»»")
        printfn "%s" (((("squeezed : length = " + (unbox<string> (padLeft (String.length ss) 2))) + ", string = «««") + ss) + "»»»")
        printfn "%s" ""
        j <- j + 1
    i <- i + 1
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)
